include/zephyr/irq.h

```diff
@@ -309,6 +309,16 @@ void z_smp_global_unlock(unsigned int key);
  */
 #define irq_is_enabled(irq) arch_irq_is_enabled(irq)
 
+/**
+ * @brief Set IRQ affinity.
+ *
+ * This routine configure a set of CPU cores that can service the @a irq.
+ *
+ * @param irq IRQ line.
+ * @param mask CPU bit mask.
+ */
+#define irq_set_affinity(irq, mask) arch_irq_set_affinity(irq, mask)
+
```

Comment:
Just a comment really, as I'm not sure what the convention is here, but would it be smart to create `CONFIG_ARCH_HAS_IRQ_AFFINITY`? Non-user-configurable, of course.

As an alternative to a Kconfig option, technically, it could even be on a per-cpu basis if it were a devicetree property (which might help for heterogeneous systems like the big+small combinations present on many ARM SoCs).

The primary reason I think that would be helpful is for testing; the testsuite could simply use that property as a filter, and then setting IRQ affinity could be very easily testable.

The secondary reason is that using a conditional declaration could alleviate the need to implement stubs everywhere (although maybe that is the convention).

What do you think?

Reply:
Thanks for this very good feedback!

> Just a comment really, as I'm not sure what the convention is here, but would it be smart to create `CONFIG_ARCH_HAS_IRQ_AFFINITY`? Non-user-configurable, of course.

I thought about doing something like this initially:

```c
#ifdef CONFIG_ARCH_HAS_IRQ_AFFINITY
#define irq_set_affinity(irq, mask) arch_irq_set_affinity(irq, mask)
#else
#define irq_set_affinity(irq, mask)
#endif /* CONFIG_ARCH_HAS_IRQ_AFFINITY */
```

but I didn't find any example of unimplemented IRQ arch APIs so I went with the current implementation (empty stub)

> The primary reason I think that would be helpful is for testing; the testsuite could simply use that property as a filter, and then setting IRQ affinity could be very easily testable.

This is a very compelling one

> As an alternative to a Kconfig option, technically, it could even be on a per-cpu basis if it were a devicetree property (which might help for heterogeneous systems like the big+small combinations present on many ARM SoCs).

Any pointers/examples?

Reply:
Let's see what others think. @npitre,  @carlocaione, @ceolin, @dcpleung 

It's possible that some architectures might not support IRQ pinning / affinity, but I also appreciate having a consistent arch interface.

Reply:
I guess as long as if can be disabled, this should be fine.

Reply:
Well... one thing to think about is that sometimes `arch_irq_*` are being implemented in the interrupt controller drivers. So in the arch layer, maybe follow the convention of `z_<arch>_irq_*` and the actual `arch_irq_*` in whichever files they need to be?

Reply:
> So in the arch layer, maybe follow the convention of z_<arch>_irq_* and the actual arch_irq_* in whichever files they need to be?

Not sure if I understood you correctly - this PR currently declare empty `arch_irq_set_affinity()` stubs according to the convention of the way each arch implements the others existing IRQ APIs, so that they can "fill in the blanks" later if they chose to implement this API, otherwise calling this API would just do nothing.

Anything further than that seems to highly arch-specific, where some of the arch simply redirects to another function (i.e. `arm_gic_irq_is_enabled()`), while some arch implemented it in place (i.e. `arch/mips/core/irq_manage.c`) or in an interrupt controller driver (i.e. `intc_swerv_pic.c`). Making these APIs consistent is probably out of the scope of this PR.

The other approach at #77828 will create an arch-level Kconfig, which will compile the `irq_set_affinity()` to nothing if it isn't implemented/enabled.

Reply:
If `arch_irq_set_affinity()` is implemented in the interrupt driver, they can then call the architecture `z_<arch>_irq_set_affinity()` (for example) if needed.

Though, ignoring the need for `z_<arch>_irq_*`, the introduction of `arch_irq_set_affinity()` needs to accommodate different implementation styles where it can be implemented in the interrupt drivers and not just inside the architecture layer.

Reply:
> If `arch_irq_set_affinity()` is implemented in the interrupt driver, they can then call the architecture `z_<arch>_irq_set_affinity()` (for example) if needed.

I don't get this - why would they want to call the architecture `z_<arch>_irq_set_affinity()`, if `arch_irq_set_affinity()` is implemented in the driver?

> the introduction of `arch_irq_set_affinity()` needs to accommodate different implementation styles where it can be implemented in the interrupt drivers and not just inside the architecture layer.

Do you mean we should allow `arch_irq_set_affinity()` to be implemented/overridden anywhere? I guess this is basically #77828?

Reply:
The `z_<arch>_irq_*` is simply where some architectures decide to do IRQ managements at arch level code. Setting affinity may or may not be needed at arch code.

#77828 would definitely be a cleaner version, and would allow driver to implement the function to set affinity.